### PR TITLE
Feature: Arachne kill timer

### PR DESCRIPTION
**Pull request by MTOnline69**
## What
Shows how long it took to kill Arachne after the fight ends. Enabled by default.

This is an improved version of [DulkirMod's one](https://github.com/inglettronald/DulkirMod/blob/master/src/main/kotlin/dulkirmod/features/ArachneTimer.kt), which has the following issues:
- The message doesn't send at all if you have bits available.
- The message often sends massively incorrect times, e.g. when:
  - You switch servers.
  - You damage Arachne but aren't in the Sanctuary when it dies.
  - You go in and out of the Sanctuary.
- The timer ends slightly after Arachne dies.
- Ugly DulkirMod Prefix

Has a slight dependency on repoPatterns in [ArachneChatMessageHider.kt](https://github.com/hannibal002/SkyHanni/blob/beta/src/main/java/at/hannibal2/skyhanni/features/chat/ArachneChatMessageHider.kt)
<details>
<summary>Images</summary>

image

</details>

## Changelog New Features
+ Added Arachne Kill Timer. - MTOnline
    * Shows how long it took to kill Arachne when the fight ends.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/chat/ArachneChatMessageHider.kt
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/chat/ArachneChatMessageHider.kt:1]: This should be made into an object so other classes can access its methods

File: src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import at.hannibal2.skyhanni.utils.ChatUtils
+import at.hannibal2.skyhanni.utils.LorenzUtils.isInIsland
+import at.hannibal2.skyhanni.utils.RegexUtils.matches
+import at.hannibal2.skyhanni.utils.SimpleTimeMark
+import at.hannibal2.skyhanni.utils.TimeUtils.format
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.time.Duration
+
+class ArachneKillTimer {
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt]: This should also be an object, and should have the @SkyHanniModule annotation to auto load it (then you remove the load module from SkyHanniMod.kt)

File: src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        if (!IslandType.SPIDER_DEN.isInIsland()) {
+            arachneSpawnedTime = SimpleTimeMark.farPast()
+        }
+    }
+
+    fun isEnabled() =
+        IslandType.SPIDER_DEN.isInIsland() && config.showArachneKillTimer
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt]:
```suggestion
    fun isEnabled() = IslandType.SPIDER_DEN.isInIsland() && config.showArachneKillTimer
```

File: src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            arachneKillTime = Duration.ZERO
+            arachneSpawnedTime = SimpleTimeMark.farPast()
+        }
+    }
+
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        if (!IslandType.SPIDER_DEN.isInIsland()) {
+            arachneSpawnedTime = SimpleTimeMark.farPast()
+        }
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt]: you should just set it to farPast here regardless of the new island

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
just a few minor things I noticed

**Review by CalMWolfs — APPROVED** [review]
worked fine, pushed a bit of code cleanup and made the text blue so that it is a bit easier to see

image

File: src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        "crystal.spawned",
+        "§c\\[BOSS] Arachne§r§f: With your sacrifice."
+    )
+    private val arachneDeathPattern by patternGroup.pattern(
+        "dead",
+        "§f.*§r§6§lARACHNE DOWN!"
+    )
+    private val arachneDamagePattern by patternGroup.pattern(
+        "damage",
+        "§f +§r§eYour Damage: §r§a[0-9,]+ §r§7\\(Position #[0-9,]+\\)"
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ArachneKillTimer.kt:51]: please use regex tests everywhre, especially for this one here

**Review by hannibal002 — CHANGES_REQUESTED** [review]
The feature works as expected. Please add regex tests for easier maintaining the class in the future.

**Review by hannibal002 — APPROVED** [review]
GitHub is perfect

image

**Pull request merged**